### PR TITLE
215 classfrequency wrapper

### PR DESCRIPTION
**Pull request by AlexanderBenz**

### CODE REVIEW

File: skactiveml/classifier/_wrapper.py [hunk 1]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
         else:
             return getattr(self.estimator, item)
+
+
+class KernelFrequencyClassifier(ClassFrequencyEstimator):
+    """KernelFrequencyClassifier
+
+    Implementation of a wrapper class for skactiveml classifiers such that
+    `predict_freq` can be used by calculating the frequencies using other 
+    models (e.g., KernelDensity from scikit-learn or ParzenWindowClassifier).
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:360]: rm (...)

File: skactiveml/classifier/_wrapper.py [hunk 2]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+
+    Parameters
+    ----------
+    estimator : sklearn.base.SkactivemlClassifier with predict_proba method
+        scikit-learn classifier that is able to deal with missing labels.
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:365]: edit, why predict_proba? rephrase

File: skactiveml/classifier/_wrapper.py (hunk 1 above)
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:360]: rm (...)

File: skactiveml/classifier/_wrapper.py [hunk 3]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        `classes` is not none.
+    metric : str or callable, default='rbf'
+        The metric must a be a valid kernel defined by the function
+        `sklearn.metrics.pairwise.pairwise_kernels`.
+    n_neighbors : int or None, default=None
+        Number of nearest neighbours. Default is None, which means all
+        available samples are considered.
+    metric_dict : dict,
+        Any further parameters are passed directly to the kernel function.
+    frequency_max_fit_len: int, default=None,
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:383]: rename to `frequency_max_labeled_samples`

File: skactiveml/classifier/_wrapper.py [hunk 4]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        Value to represent the frequency estimator sliding window size for X,
+        y and sample weight. If 'None' the windows is unrestricted in size.
+    frequency_replacing_strategy: str, default='last'
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:386]: rename to `frequency_replacement_method`

File: skactiveml/classifier/_wrapper.py [hunk 5]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        -------
+        F: array-like of shape (n_samples, classes)
+            The class frequency estimates of the input samples. Classes are
+            ordered according to `classes_`.
+        """
+        check_is_fitted(self)
+        X = check_array(X)
+
+        # Predict zeros because of missing training data.
+        if not self.is_fitted_:
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:566]: check why this is necessary as before we already have `check_is_fitted`.

File: skactiveml/classifier/_wrapper.py [hunk 6]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+            must have the same shape as y.
+        fit_kwargs : dict-like
+            Further parameters as input to the 'fit' method of the 'estimator'.
+
+        Returns
+        -------
+        self : KernelFrequencyClassifier,
+            The KernelFrequencyClassifier is fitted on the training data.
+        """
+        self._fit_frequency_estimator(
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:492]: high potential for redundancies. maybe start with estimator.fit() and then access ExtLabelEnc and classes from estimator for determining the frequencies. Thereby, we do not need to check if classes and labelencoder are similar

File: skactiveml/classifier/_wrapper.py [hunk 7]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+                "ensure_min_features": 0,
+                "ensure_2d": False,
+                "force_all_finite": False,
+                "dtype": None,
+            }
+
+        # Check common classifier parameters.
+        check_classifier_params(
+            self.classes, self.missing_label, self.cost_matrix
+        )
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:666]: check if missing labels are similar: self, self.estimator

File: skactiveml/classifier/_wrapper.py [hunk 8]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+
+        # Store and check random state.
+        self.random_state_ = check_random_state(self.random_state)
+
+        # Create label encoder.
+        self._le = ExtLabelEncoder(
+            classes=self.classes, missing_label=self.missing_label
+        )
+
+        # Check input parameters.
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:676]: remove redundancies

File: skactiveml/classifier/_wrapper.py [hunk 9]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        # Check common classifier parameters.
+        check_classifier_params(
+            self.classes, self.missing_label, self.cost_matrix
+        )
+
+        # Store and check random state.
+        self.random_state_ = check_random_state(self.random_state)
+
+        # Create label encoder.
+        self._le = ExtLabelEncoder(
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:672]: necessary?

File: skactiveml/classifier/_wrapper.py [hunk 10]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+                for c in range(len(self._le.classes_))
+            ]
+        else:
+            self._le.fit_transform(self.classes)
+            check_X_dict["ensure_2d"] = False
+        X = check_array(X, **check_X_dict)
+        check_consistent_length(X, y)
+
+        # Update detected classes.
+        self.classes_ = self._le.classes_
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:702]: get from estimator

File: skactiveml/classifier/_wrapper.py [hunk 11]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        scikit-learn classifier that is able to deal with missing labels.
+    classes : array-like of shape (n_classes,), default=None
+        Holds the label for each class. If none, the classes are determined
+        during the fit.
+    missing_label : scalar or string or np.nan or None, default=np.nan
+        Value to represent a missing label.
+    cost_matrix : array-like of shape (n_classes, n_classes)
+        Cost matrix with `cost_matrix[i,j]` indicating cost of predicting class
+        `classes[j]` for a sample of class `classes[i]`. Can be only set, if
+        `classes` is not none.
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:374]: add `use_only_marginal_frequencies` (evlt rename): sums up frequencies as the number of nearby labels and multiplies it by predict_proba from estimator.

File: skactiveml/classifier/_wrapper.py [hunk 12]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+    metric_dict : dict,
+        Any further parameters are passed directly to the kernel function.
+    frequency_max_fit_len: int, default=None,
+        Value to represent the frequency estimator sliding window size for X,
+        y and sample weight. If 'None' the windows is unrestricted in size.
+    frequency_replacing_strategy: str, default='last'
+        Defines how old instances of the frequency estimator are replaced.
+        'last': First in First out
+        'random' : replacing old samples randomly
+    class_prior : float or array-like, shape (n_classes), default=0
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:390]: remove (prior is not included in frequencies)

File: skactiveml/classifier/_wrapper.py [hunk 13]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+    Parameters
+    ----------
+    estimator : sklearn.base.SkactivemlClassifier
+        The wrapped classifier.
+    classes : array-like of shape (n_classes,), default=None
+        Holds the label for each class. If none, the classes are determined
+        during the fit.
+    missing_label : scalar or string or np.nan or None, default=np.nan
+        Value to represent a missing label.
+    cost_matrix : array-like of shape (n_classes, n_classes)
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:802]: redundant. similar problem with sklearn-wrapper? what if sklearn clf already supports cost_matrix?

File: skactiveml/classifier/_wrapper.py [hunk 14]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        Cost matrix with `cost_matrix[i,j]` indicating cost of predicting class
+        `classes[j]` for a sample of class `classes[i]`. Can be only set, if
+        `classes` is not none.
+    max_fit_len: int, default=None,
+        Value to represent the estimator sliding window size for X, y and
+        sample weight. If 'None' the windows is unrestricted in size.
+    replacing_strategy: str, default='last'
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:809]: `replacement_method`

File: skactiveml/classifier/_wrapper.py [hunk 15]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+    classes : array-like of shape (n_classes,), default=None
+        Holds the label for each class. If none, the classes are determined
+        during the fit.
+    missing_label : scalar or string or np.nan or None, default=np.nan
+        Value to represent a missing label.
+    cost_matrix : array-like of shape (n_classes, n_classes)
+        Cost matrix with `cost_matrix[i,j]` indicating cost of predicting class
+        `classes[j]` for a sample of class `classes[i]`. Can be only set, if
+        `classes` is not none.
+    max_fit_len: int, default=None,
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:806]: `subsample_size`

File: skactiveml/classifier/_wrapper.py [hunk 16]
```diff
@@ -322,3 +350,874 @@ def __getattr__(self, item):
...
+        return X, y, sample_weight
+
+    def __getattr__(self, item):
+        if "estimator_" in self.__dict__ and hasattr(self.estimator_, item):
+            return getattr(self.estimator_, item)
+        else:
+            raise AttributeError(f"{item} does not exist")
+
+
+class SubSampleEstimator(SkactivemlClassifier, MetaEstimatorMixin):
```
Comment (dakot) [inline skactiveml/classifier/_wrapper.py:784]: `SubsampleEstimator`

File: skactiveml/classifier/_wrapper.py (hunk 13 above)
Comment (dakot) [reply skactiveml/classifier/_wrapper.py:802]: at minimum: ensure that everything is the same

**Review by dakot — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/_wrapper.py (hunk 11 above)
Comment (tpham93) [reply skactiveml/classifier/_wrapper.py]: `use_estimator_proba`

**Review by dakot — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/_wrapper.py [hunk 17]
```diff
@@ -322,3 +342,800 @@ def __getattr__(self, item):
...
+        class_frequency_estimator is not None.
+    random_state : int or RandomState instance or None, default=None
+        Determines random number for 'predict' method. Pass an int for
+        reproducible results across multiple method calls.
+    """
+
+    def __init__(
+        self,
+        estimator,
+        class_frequency_estimator=None,
```
Comment (tpham93) [inline skactiveml/classifier/_wrapper.py:394]: Remove default argument

**Review by tpham93 — CHANGES_REQUESTED** [review]
